### PR TITLE
Switch extension downloader and extensions feed to use Guzzle HTTP library

### PR DESCRIPTION
**Pull request by mattwire**
Overview
----------------------------------------
Switch extension download and extensions feed to use Guzzle HTTP library.

This fixes some situations (eg. shared servers with open_basedir restrictions in effect) where zip file was not downloaded properly and extension could not be updated.
It also means that we are no longer using our custom `CRM_Utils_HttpClient::fetch()` function from within civicrm core.

Before
----------------------------------------
Using civicrm core function `CRM_Utils_HttpClient::fetch()` which uses cURL internally.

After
----------------------------------------
Using guzzle HTTP library which has been shipped with core for a long time now.

Fixes some situations (eg. shared servers with open_basedir restrictions in effect) where zip file was not downloaded properly and extension could not be updated.

Technical Details
----------------------------------------
Simple change, following existing code examples within civicrm core.

Comments
----------------------------------------

### CODE REVIEW

**Comment by mattwire** [comment]
@seamuslee001 test failures seem more like a test environment issue?

**Comment by seamuslee001** [comment]
@mattwire seems like it, let me get into it with Tim n Mathieu

**Comment by seamuslee001** [comment]
@mattwire I found what seems to be the problem which is this https://github.com/guzzle/guzzle/blob/master/src/Handler/CurlFactory.php#L231 by default Guzzle with CURL only allows `http://`  and `https://` you can see the definition here https://www.php.net/manual/en/function.curl-setopt.php and the thing is that is using a `file://` URL https://github.com/civicrm/civicrm-core/blob/master/tests/phpunit/api/v3/ExtensionTest.php#L26 When i enabled the FILE protocol locally, i ran into a different error but haven't been able to track that down yet.

**Comment by mattwire** [comment]
@seamuslee001 Thanks for digging into that. So seems like a test only issue because we're unlikely to be using a file:// repo in the real world. @eileenmcnaughton any ideas how we could modify this test setup to work better with guzzle? I know you always say guzzle has better support for mocking etc.

**Comment by eileenmcnaughton** [comment]
@mattwire I think you want to fix that class to use the `GuzzleTestTrait`

I'd check your assumption " seems like a test only issue because we're unlikely to be using a file:// repo in the real world" with @totten though as he added the test & put 'file://' in there deliberately so may see a real world use case

**Comment by agileware-justin** [comment]
Good work @mattwire

**Comment by mattwire** [comment]
test this please

**Comment by demeritcowboy** [comment]
This has conflicts now.
I'm not sure which test was failing. I can't personally think of a real-world reason.

**Comment by mattwire** [comment]
test this please

**Comment by mattwire** [comment]
test this please

File: CRM/Extension/Downloader.php [hunk 1]
```diff
@@ -75,11 +95,9 @@ public function checkRequirements($extensionInfo = NULL) {
       $requiredExtensions = CRM_Extension_System::singleton()->getManager()->findInstallRequirements([$extensionInfo->key], $extensionInfo);
       foreach ($requiredExtensions as $extension) {
         if (CRM_Extension_System::singleton()->getManager()->getStatus($extension) !== CRM_Extension_Manager::STATUS_INSTALLED && $extension !== $extensionInfo->key) {
-          $requiredExtensionInfo = CRM_Extension_System::singleton()->getBrowser()->getExtension($extension);
-          $requiredExtensionInfoName = empty($requiredExtensionInfo->name) ? $extension : $requiredExtensionInfo->name;
           $errors[] = [
             'title' => ts('Missing Requirement: %1', [1 => $extension]),
-            'message' => ts('You will not be able to install/upgrade %1 until you have installed the %2 extension.', [1 => $extensionInfo->name, 2 => $requiredExtensionInfoName]),
+            'message' => ts('You will not be able to install/upgrade %1 until you have installed the %2 extension.', [1 => $extensionInfo->name, 2 => $extension]),
```
Comment (demeritcowboy) [inline CRM/Extension/Downloader.php:100]: This undoes https://github.com/civicrm/civicrm-core/pull/21852 - just wondering if it's on purpose or a rebase/conflict resolution mixup?

Comment (mattwire) [reply CRM/Extension/Downloader.php]: @demeritcowboy Gone! thanks that was a rebase/mixup

**Comment by totten** [comment]
> @mattwire I'm pretty sure that file:/// was only used in the test because it was easier than trying to access a webserver from the test. 

Yeah, it was probably the easiest way to mock the HTTP call.

Of course, with Guzzle, we can do another style of mocking... Added f4a52966bfc24453fea03753259c4b698f9b870d.

+1 for merge (as long as someone has `r-run`)

**Comment by demeritcowboy** [comment]
Have run it. Everything seems fine. Downloading/installing all works. It sees that updates are available.

**Comment by mattwire** [comment]
Thankyou :-)

**Pull request merged**